### PR TITLE
C++20 Support

### PR DESCRIPTION
**Pull request by swbsi**
There are also changes in internal and bsicommon (for mobile builds).

imodel-native-internal: https://github.com/iTwin/imodel-native-internal/tree/cxx20

### CODE REVIEW

File: iModelCore/libsrc/csmap/Include/cs_EpsgStuff.h [hunk 1]
```diff
@@ -344,6 +344,7 @@ class TcsEpsgCode
 	bool operator< (const std::wstring& epsgCode) const;
 	bool operator== (unsigned long epsgCode) const;
 	bool operator== (unsigned int epsgCode) const;
+	bool operator== (const TcsEpsgCode& rhs) const; // to avoid comparison term reordering issues in c++20
```
[the remark this answers is not in the document]
Comment (swbsi) [reply iModelCore/libsrc/csmap/Include/cs_EpsgStuff.h:347]: This is a pretty good explanation: https://brevzin.github.io/c++/2019/07/28/comparisons-cpp20/

To summarize, primaries can now be reversed and secondaries can now be rewritten in terms of primaries.

**Review by khanaffan — APPROVED** [review]

File: iModelCore/ecobjects/test/NonPublished/ECInstanceJsonWriterTests.cpp [hunk 2]
```diff
@@ -441,46 +441,46 @@ TEST_F(ECInstanceJsonWriterTests, WriteEmbeddedStructValueTest)
                  }
               ]
            }
-        })");
+        })"));
 
     // Act
 
-    BeJsDocument actualJsonValue1;
+    BeJsDocument actualJsonValue1;
```
Comment (ColinKerr) [inline iModelCore/ecobjects/test/NonPublished/ECInstanceJsonWriterTests.cpp:448]: Are there some unexpected or unintended line ending changes happening?  There are a lot of seemingly unchanged changed lines related to json.  Why are they being flagged as changed and is it intentional?

Comment (swbsi) [reply iModelCore/ecobjects/test/NonPublished/ECInstanceJsonWriterTests.cpp:448]: Not sure. I will check…

**Comment by swbsi** [comment]
FYI, I realize it’s not building. I forgot to make c++20 the default in the make files…I’ve been building with the environment variable set.

File: iModelCore/libsrc/facebook/VendorApi/folly/MicroSpinLock.h [hunk 3]
```diff
@@ -40,19 +40,19 @@
 #include <array>
 #include <cinttypes>
 #include <type_traits>
-#if defined (BENTLEY_CHANGE)
+#if defined (BENTLEY_CHANGE)
```
Comment (ColinKerr) [inline iModelCore/libsrc/facebook/VendorApi/folly/MicroSpinLock.h:43]: more line ending issues?

File: iModelCore/ecobjects/test/NonPublished/ECInstanceJsonWriterTests.cpp (hunk 2 above)
Comment (ColinKerr) [reply iModelCore/ecobjects/test/NonPublished/ECInstanceJsonWriterTests.cpp]: looks good

**Review by grigasp — APPROVED** [review]

**Review by rschili — APPROVED** [review]

File: iModelCore/Bentley/PublicAPI/Bentley/WString.h [hunk 4]
```diff
@@ -542,4 +542,19 @@ typedef T_Utf8StringVector*         T_Utf8StringVectorP, &T_Utf8StringVectorR;
...
+    Utf8Char m_characters[N];
+    
+    template<size_t... I>
+    constexpr Utf8Chars(const char8_t (&input)[N], std::index_sequence<I...>) : m_characters {static_cast<Utf8Char>(input[I])...} {}
+    
+public:
+    constexpr Utf8Chars(const char8_t (&input)[N]) : Utf8Chars(input, std::make_index_sequence<N>()) {}
+
+    constexpr operator Utf8CP() const { return m_characters; }
+    constexpr Utf8CP operator&() const { return m_characters; }
```
Comment (pmconne) [inline iModelCore/Bentley/PublicAPI/Bentley/WString.h:557]: Yuck!

**Review by pmconne — APPROVED** [review]

File: iModelCore/Bentley/PublicAPI/Bentley/WString.h (hunk 4 above)
Comment (swbsi) [reply iModelCore/Bentley/PublicAPI/Bentley/WString.h:557]: I agree. Supposedly, wonderful new utf8 APIs are coming in future C++ versions. However, for now, we are stuck doing this. They made char8_t a completely separate type in C++20...

**Review by kabentley — APPROVED** [review]

**Pull request merged**